OSDOCS-3780&3781&3782:BYOK encryption (#55014)
openshift/openshift-docs: modules/persistent-storage-byok.adoc

// Module included in the following assemblies:
//
// storage/container_storage_interface/persistent-storage-csi-azure.adoc
// storage/container_storage_interface/persistent-storage-csi-ebs.adoc
// storage/container_storage_interface/persistent-storage-csi-gcp-pd.adoc

:_content-type: CONCEPT
[id="byok_{context}"]
= User-managed encryption

> The user-managed encryption feature allows you to provide keys during installation that encrypt {product-title} node root volumes, and enables the default storage class to use these keys to encrypt provisioned storage volumes. You specify the custom key in the `platform.<cloud_type>.defaultMachinePlatform` field in the install-config YAML file.

Reviewer: Suggest to change `You specify the custom key` to:
`You must specify the custom key`, as this is the only place that enables the default storage class to be encrypted. Specify the key in other field won't work for this.

Reviewer 2: +1 EBS operator creates two storage class, replaced default storage class by "managed storage class" or "storage class managed by the storage operator".

This features supports the following storage types:

* Amazon Web Services (AWS) Elastic Block storage (EBS)

* Microsoft Azure Disk storage

* Google Cloud Platform (GCP) persistent disk (PD) storage

> // Procedure on how to make this happen. Azure is covered in Azure assembly. AWS and GCP???

Reviewer 3: This comment looks like it could be removed now

Author: Yes. Done.

// Module included in the following assemblies:
//
// storage/container_storage_interface/persistent-storage-csi-azure.adoc
// storage/container_storage_interface/persistent-storage-csi-ebs.adoc
// storage/container_storage_interface/persistent-storage-csi-gcp-pd.adoc

:_content-type: CONCEPT
[id="byok_{context}"]
= About "bring your own key" encryption

> The "Bring your own key" (BYOK) feature encrypts root volumes and the default storage class.

Reviewer 2: "openshift node root volumes" just to clarify what we mean by root volumes.

Reviewer 2: I am not sure we should keep the BOYK terminology, I think we should align with the installer doc and use "user managed encryption key" instead. Thoughts?

Reviewer 4: +1 words like "Customer-managed(user-managed) encryption keys" is better than "BOYK".

> The "Bring your own key" (BYOK) feature encrypts root volumes and the default storage class.

Reviewer 4: For this `encrypts root volumes and the default storage class.` and  `The installation program also encrypts the default storage class with this key` in Line 13:
Here I think it doesn't encrypt the storageclass, but encrypt the PersistentVolume (created with pre-created storageclass by CSI Driver Operator)

> When starting up managed {product-title} clusters, you can pass a custom Key Management Service (KMS) key to the installation program to encrypt the root volumes for the nodes in the cluster. The installation program also encrypts the default storage class with this key. This allows you to encrypt storage volumes with your selected key, instead of the platform's default account key.

Reviewer: Based on https://issues.redhat.com/browse/CORS-2563?focusedCommentId=21842953&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-21842953, 
```
The installation program also encrypts the default storage class with this key
``` 
only happens when the custom key was specified in platform.{cloud_type}.defaultMachinePlatform field in the install-config.yaml. This should be highlighted in this section.

// Module included in the following assemblies:
//
// storage/container_storage_interface/persistent-storage-csi-azure.adoc
// storage/container_storage_interface/persistent-storage-csi-ebs.adoc
// storage/container_storage_interface/persistent-storage-csi-gcp-pd.adoc

:_content-type: CONCEPT
[id="byok_{context}"]
= About "bring your own key" encryption

// TP or GA???

The "Bring your own key" (BYOK) feature encrypts root volumes and the default storage class.

> When starting up managed {product-title} clusters, you can pass a custom Key Management Service (KMS) key to the installer, and the installer and `cluster-storage-operator` encrypts the root volumes for the nodes in the cluster, and also applies this key to encrypt the default storage class.

Reviewer 5: ```suggestion
When starting up managed {product-title} clusters, you can pass a custom Key Management Service (KMS) key to the installation program to encrypt the root volumes for the nodes in the cluster. The installation program also encrypts the default storage class with this key.
```
Rewording/minimalism suggestion, as well as replacing "installer" with "installation program" per style guide. Up to you whether you want to specifically mention the `cluster-storage-operator` as doing the encryption.

> This allows you to encrypt your owned volumes with your selected key, instead of the platform's default account key.

Reviewer 5: ```suggestion
This allows you to encrypt storage volumes with your selected key, instead of the platform's default account key.
```
the phrase "your owned volumes" seemed strange to me, but feel free to leave it if you prefer it.

This features supports the following storage types:

* Amazon Web Services (AWS) Elastic Block storage (EBS)

* Microsoft Azure Disk storage

* Google Cloud Platform (GCP) persistent disk (PD) storage

> // Procedure on how to make this happen???

Reviewer 5: I wrote a procedure for this for Azure:
https://docs.openshift.com/container-platform/4.12/installing/installing_azure/enabling-user-managed-encryption-azure.html
With an accompanying module inside the installation assemblies for finalizing:
https://docs.openshift.com/container-platform/4.12/installing/installing_azure/installing-azure-network-customizations.html#finalizing-encryption_installing-azure-network-customizations

Not sure about AWS and GCP though. @jhixson74 Do you know if we need to update the "Enabling user managed encryption" page with any new procedure or configuration to describe this feature?

Reviewer 5: I also have https://issues.redhat.com/browse/CORS-2080 which I believe is covering the installer portion of extending the azure disk encryption to the default storage class
